### PR TITLE
Intersection observer

### PR DESCRIPTION
**Pull request by VladDubrovskis**
Todo:
- [x] Wait for release of Polyfill service
- [x] Sort out the browser tests for lazy loading as currently only runs half of assertions
- [x] update the lazyLoading config to be an object of options
- [x] Update `origami.json` to include the required polyfills
- [x] Add polyfills on the demos
- [x] Update `README`

### CODE REVIEW

File: demos/Individual-Ad-Lazy-Load-Margin.html [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<!DOCTYPE html>
+<html class="o-hoverable-on ">
+<head>
+	<meta charset="utf-8" />
+	<meta http-equiv="X-UA-Compatible" content="IE=Edge" />
+	<title>o-ads: Individual-Ad-With-Out-Of-Page demo</title>
+	<meta name="viewport" content="initial-scale=1.0, width=device-width" />
+	<script src="//polyfill.webservices.ft.com/v1/polyfill.min.js?features=default"></script>
```
Comment (adgad) [inline demos/Individual-Ad-Lazy-Load-Margin.html]: We should add the polyfill here

File: demos/Individual-Ad-Lazy-Load-Margin.html [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	<link rel="stylesheet" href="/bundles/css?modules=o-ads:/demos/src/demo.scss" />
+</head>
+<body>
+<script data-o-ads-config type="application/json">
+	{
+		"gpt": {
+			"network": 5887,
+			"site": "test.5887.origami"
+		},
+		"lazyLoad": true,
```
Comment (adgad) [inline demos/Individual-Ad-Lazy-Load-Margin.html]: I wonder if this would be better as an object e.g. 
`lazyLoad: { margin: "1000%" }`

**Comment by adgad** [comment]
We should also update the browserFeatures part of origami.json with this and anything we missed last time (e.g. Array.from) http://origami.ft.com/docs/syntax/origamijson/

File: demos/Individual-Ad-Lazy-Load-Margin.html (hunk 2 above)
Comment (VladDubrovskis) [inline demos/Individual-Ad-Lazy-Load-Margin.html]: good idea!

File: README.md [hunk 3]
```diff
@@ -107,6 +107,34 @@ Parameter | Description
...
 
+## Events
+### `oAds.adServerLoadError`
+Triggered if the library fails to load external JS GPT library, meaning no advertising will work. Can be used if you wish to have a fallback when you know the adverts will not display.
+
+### `oAds.ready`
+Slot has been inited in the oAds library and is about to be requested from the ad server (deferred if lazy loading is on)
+
+### `oAds.rendered`
+Triggered once the as has been rendered on the page.
```
Comment (leggsimon) [inline README.md]: "once the **ad**"

File: docs/SLOT_CONFIG_REFERENCE.md [hunk 4]
```diff
@@ -134,23 +134,36 @@ To serve pop-up, pop-under, or floating creatives to your website, you’ll need
...
 Defines if the slot should be lazy loaded, e.g. only requested from the advertising server once it comes into view. Default is `false`.
+Support following properties:
+* **viewportMargin** - Sets the margin of viewport to use when deciding if the advert is in view. Useful if you want to request and display the advert just before it comes into view. Works as regular margin definitions. Make sure you always specify the dimensions with either `px` or `%`, e.g. `100% 0%`, or `100px 0px`. Default is `0%`.
+* **threshold** -  Sets threshold at which advert should be requested and displayed. Value can be anything between 0 and 100. Default is `0`, which means as soon as first pixel in in view the advert will get loaded.
+
+
+**IMPORTANT** Starting with version 5 oAds is relying on [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API). If one ins not avaailable in browser, you may need to provide a polyfill. The easiest way is to include with [Polyfill.io] on your page.
+E.g.
+```html
+<script src="https://cdn.polyfill.io/v2/polyfill.min.js?features=IntersectionObserver|always|gated"></script>
```
Comment (adgad) [inline docs/SLOT_CONFIG_REFERENCE.md]: Probably not a good idea to recommend `always`  and `gated` here?

File: src/js/slot.js [hunk 5]
```diff
@@ -2,280 +2,316 @@ const utils = require('./utils');
...
-		this.defer = true;
-		let renderSlot = function(slot) {
-			/* istanbul ignore else */
-			if(!slot.rendered) {
-				slot.fire('render');
-				slot.rendered = true;
-			}
-		}.bind(null, this);
-		utils.once('inview', renderSlot, this.container);
-		//Master/Companion ads don't work with lazy loading, so if a master ad loads trigger
```
Comment (adgad) [inline src/js/slot.js]: Worth leaving this in I think :)

**Comment by adgad** [comment]
LGTM (minor comments aside)

**Comment by JakeChampion** [comment]
🎉

**Pull request merged**